base/dates/io.jl

```diff
@@ -34,123 +34,153 @@ const MONTHTOVALUE = Dict{UTF8String,Dict{UTF8String,Int}}("english"=>english)
 const MONTHTOVALUEABBR = Dict{UTF8String,Dict{UTF8String,Int}}("english"=>abbrenglish)
 
 # Date/DateTime Parsing
-abstract Slot{P<:AbstractTime}
+abstract Slot{T<:Any}
 
-immutable DelimitedSlot{P<:AbstractTime} <: Slot{P}
-    i::Int
-    period::Type{P}
+immutable DelimitedSlot{T<:Any} <: Slot{T}
+    parser::Type{T}
+    letter::Char
     width::Int
-    option::Int
-    locale::AbstractString
+    transition::Union{Regex,AbstractString}
 end
 
-immutable FixedWidthSlot{P<:AbstractTime} <: Slot{P}
-    i::Int
-    period::Type{P}
+immutable FixedWidthSlot{T<:Any} <: Slot{T}
+    parser::Type{T}
+    letter::Char
     width::Int
-    option::Int
-    locale::AbstractString
 end
 
 immutable DateFormat
     slots::Array{Slot,1}
-    begtran # optional transition from the start of a string to the 1st slot
-    trans #trans[i] == how to transition FROM slots[i] TO slots[i+1]
+    prefix::AbstractString # optional transition from the start of a string to the 1st slot
+    locale::AbstractString
 end
 
-immutable DayOfWeekSlot <: AbstractTime end
+abstract DayOfWeekSlot
 
-duplicates(slots) = any(map(x->count(y->x.period==y.period,slots),slots) .> 1)
+# Slot rules allowed for where ismatch(r"^[a-zA-Z]$",c)
+slotrule(::Type{Val{'y'}}) = Year
+slotrule(::Type{Val{'m'}}) = Month
+slotrule(::Type{Val{'u'}}) = Month
+slotrule(::Type{Val{'U'}}) = Month
+slotrule(::Type{Val{'e'}}) = DayOfWeekSlot
+slotrule(::Type{Val{'E'}}) = DayOfWeekSlot
+slotrule(::Type{Val{'d'}}) = Day
+slotrule(::Type{Val{'H'}}) = Hour
+slotrule(::Type{Val{'M'}}) = Minute
+slotrule(::Type{Val{'S'}}) = Second
+slotrule(::Type{Val{'s'}}) = Millisecond
+slotrule{c}(::Type{Val{c}}) = Union{}
```

Comment:
I dislike the use of multiple dispatch to pun on a Dict (or any other primitive data structure for that matter). This is being used to form a simple Associative mapping, there is no need to hide that in the generic dispatch machinary. The use of `Val` is a strong indicator to me that this code is not optimal style. This is also much more verbose and less clear than using the actual data structure this is mimicking, it is substantially slower, you can't change or remove items, and extending it at runtime can cause you to run into code statepoints errors (I don't have great terminology for this error, but it's somewhat related to # 265).